cmd/clusterctl/cmd/restore.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/client"
+)
+
+type restoreOptions struct {
+	toKubeconfig        string
+	toKubeconfigContext string
+	namespace           string
+	glob                string
+}
+
+var ro = &restoreOptions{}
+
+var restoreCmd = &cobra.Command{
+	Use:   "restore",
+	Short: "Restore Cluster API objects from file by glob. Object files are searched in config directory",
+	Long: LongDesc(`
+		Restore Cluster API objects from file by glob. Object files are searched in the default config directory
+		or in the provided directory.
+
+		TODO`),
+	Example: Examples(`
+		Restore Cluster API objects from file by glob. Object files are searched in config directory.
+		clusterctl restore my-cluster`),
+	Args: cobra.NoArgs,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runRestore()
+	},
+}
+
+func init() {
+	restoreCmd.Flags().StringVar(&ro.toKubeconfig, "kubeconfig", "",
+		"Path to the kubeconfig file for the target management cluster to restore objects to. If unspecified, default discovery rules apply.")
+	restoreCmd.Flags().StringVar(&ro.toKubeconfigContext, "kubeconfig-context", "",
+		"Context to be used within the kubeconfig file for the target management cluster. If empty, current context will be used.")
+	restoreCmd.Flags().StringVarP(&ro.namespace, "namespace", "n", "",
+		"The namespace to restore Cluster API objects to in the target management cluster. If unspecified, the current context's namespace is used.")
+	restoreCmd.Flags().StringVar(&ro.glob, "glob", "",
```

Comment:
Good question. 

I think the general use case would be if a user wants to restore a certain cluster from a certain namespace. Let's look at a few examples:

 imagine I have `cluster-a` and `cluster-b` in namespace `my-namespace`. If I backup both clusters using the following command

```
clusterctl backup --namespace my-namespace
```

then both clusters in that namespace will be backed-up to the default config / objects directory. I end up with this kind of filetree (leaving out some of the other object files like `Secrets` for simplicity):

```
~/.cluster-api/objects
    | - Cluster_cluster-a_my-namespace.yaml
    | - Cluster_cluster-b_my-namespace.yaml
```

If I then want to restore _only_ `cluster-a`, I have no ability to do that unless I provide a glob. This would look like:

```
clusterctl restore --glob cluster-a
``` 

And it will only match files (and restore those objects) that match "cluster-a" as a substring.

Further, I can imagine the use case where only certain namespaces are restored. For example, if I have `cluster-foo` and `cluster-bar` in namespace `namespace-a` and `cluster-baz` in namespace `namespace-b`, I could backup both namespaces with:
```
clusterctl backup --namespace namespace-a
clusterclt backup --namespace namespace-b
```

Then, the default config objects directory would look like:

```
~/.cluster-api/objects
    | - Cluster_cluster-foo_namespace-a.yaml
    | - Cluster_cluster-bar_namespace-a.yaml
    | - Cluster_cluster-baz_namespace-b.yaml
```

Then, I could choose to restore _only_ those clusters that were in `namespace-a` with:
```
clusterctl restore --glob namespace-a
```
This could be _refined even further_ by providing both the namespace and the cluster name:
```
clusterctl restore --glob cluster-foo_namespace-a
```
will only restore objects that are associated with cluster `cluster-foo` from `namespace-a`.

This probably warrants being documented better in the help message. If this sound agreeable, I can go ahead and flesh out the help message

Reply:
Uhmm
I'm not sure the back/restore single file scenario really holds, given that the naming schema is kind-name-namespace, and so with the example above you are only to restore the cluster object, but not the entire hierarchy of objects belonging to a cluster.
WRT to restore a single namespace at time, this could make sense, but TBH I think we should reccomand to backup different namespaces in different folders.
Given that, my personal opinion is to remove this flag/parameter for the first iteration and keep the things simple, but it would be interesting to have some other opinions here.
@vincepri

Reply:
Good point - I think removing the glob option for now and requiring the directory would be a good first pass. And in the future, if we wanted some kind of partial restore or glob based restore, we can add it in.